iree/samples/static_library/CMakeLists.txt

```diff
@@ -0,0 +1,54 @@
+# Copyright 2021 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# if(${IREE_BUILD_STATIC_LIBRARY_SAMPLES} AND NOT ${IREE_BUILD_SAMPLES})
+#   return()
+# endif()
+
+if(NOT ${IREE_TARGET_BACKEND_DYLIB-LLVM-AOT}
+  OR NOT ${IREE_HAL_DRIVER_DYLIB}
+  OR NOT ${IREE_BUILD_COMPILER}
+  OR NOT ${IREE_BUILD_SAMPLES})
+  return()
+endif()
+
+iree_bytecode_module(
+  NAME
+    simple_mul
+  SRC
+    "simple_mul.mlir"
+  C_IDENTIFIER
+    "iree_samples_static_library_simple_mul"
+  FLAGS
+    "-iree-input-type=mhlo"
+    "-iree-mlir-to-vm-bytecode-module"
+    "-iree-hal-target-backends=dylib-llvm-aot"
+    "-iree-llvm-static-library-output-path=${CMAKE_CURRENT_BINARY_DIR}/simple_mul.o"
+  PUBLIC
+  TESTONLY
+)
+
+iree_cc_binary(
+  NAME
+    static_library_demo
+  SRCS
+    "static_library_demo.c"
+  DEPS
+    ::simple_mul_c
+    iree::runtime
+    iree::base
+    iree::hal
+    iree::hal::local
+    iree::hal::local::loaders::static_library_loader
+    iree::hal::local::task_driver
+    iree::modules::hal
+    iree::task::api
+    iree::vm
+    iree::vm::bytecode_module
+    ${CMAKE_CURRENT_BINARY_DIR}/simple_mul.o
```

Comment:
If this generated file isn't getting found ([logs](https://source.cloud.google.com/results/invocations/1e6d073a-85f1-4e30-b139-50126d000de7/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Fx86-swiftshader%2Fpresubmit/log)), 

```
ninja: error: 'iree/samples/static_library/simple_mul.o', needed by 'iree/samples/static_library/static_library_demo', missing and no known rule to make it
```

maybe try something like 
https://github.com/google/iree/blob/4d3b33a0f3e9a5d9266463efbafb6cb2be9e3251/iree/base/testing/CMakeLists.txt#L7-L23
or declare that the `iree_bytecode_module` rule generates `simple_mul.o`

(@GMNGeoffrey any suggestions?)

Reply:
Discussed offline a bit. We found https://stackoverflow.com/a/14873818 helpful here.

pseudocode that expresses our intent to CMake directly:

```cmake
# (setup args for iree-translate)

# custom command for iree-translate
add_custom_command(
  OUTPUT
    "simple_mul.vmfb"
    "simple_mul.o"
  COMMAND ${_TRANSLATE_TOOL_EXECUTABLE} ${_ARGS}
  DEPENDS iree_tools_iree-translate
)

# tell CMake not to compile the generated file
SET_SOURCE_FILES_PROPERTIES(
  simple_mul.o
  PROPERTIES
  EXTERNAL_OBJECT true
  GENERATED true
)

# tell CMake how to link the generated file
SET_TARGET_PROPERTIES(
  template
  PROPERTIES
  LINKER_LANGUAGE C 
)

# (setup args for generate_embed_data)

# custom command for generate_embed_data
add_custom_command(
  OUTPUT
    "simple_mul_c.h"
    "simple_mul_c.c"
  COMMAND generate_embed_data ${_ARGS}
  DEPENDS generate_embed_data simple_mul.vmfb
)
```

Using the raw cmake here will make it easier to switch from .vmfb -> c_embed_data to emit_c. We could also fold some of this into our existing rules (`iree_bytecode_module` or a fork of it). That would help with reuse if other places want to use this feature, but it will take a bit to figure out the right structure.